server.js

```diff
@@ -45,53 +71,94 @@ function runClient(client) {
         socket.emit('upvote', answers[upvotedId][upvotedNumber]);
       }
     }
+
+    // clearing answers
+    if (answersInQueuePerIP[ipAddress]) {
+      answersInQueuePerIP[ipAddress] -= answers[client.id] ? answers[client.id].length : 0;
+    }
     delete answers[client.id];
     socket.emit('remove answers', client.id);
+
+    // clearing questions
+    if (questions[client.id] && questionsInQueuePerIP[ipAddress]) {
+      questionsInQueuePerIP[ipAddress]--;
+    }
+    if (currentQuestion.id === client.id && questionsInQueuePerIP[ipAddress]) {
+      questionsInQueuePerIP[ipAddress]--;
+    }
+    delete questions[client.id];
+
+    // clearing user
     delete clients[client.id];
     socket.emit('clients online', Object.keys(clients).length);
+
+    // clearing ip-wide data
+    if (answersInQueuePerIP[ipAddress] === 0) {
+      delete answersInQueuePerIP[ipAddress];
+    }
+    if (questionsInQueuePerIP[ipAddress] === 0) {
+      delete questionsInQueuePerIP[ipAddress];
+    }
+
+    log('clearing on disconnect');
   });
 
   client.on('submit question', function(data) {
-    data = data.replace(/</g, "&lt;").replace(/>/g, "&gt;");
-    console.log('question received: ' + data);
+    data = escapeHtml(data);
+    questionsInQueuePerIP[ipAddress] = questionsInQueuePerIP[ipAddress] || 0;
+    // log('question received: ' + data);
     if (client.id in questions) {
-      console.log('client already has question in queue');
+      log('client already has question in queue');
+      client.emit('max questions');
+    } else if (questionsInQueuePerIP[ipAddress] >= maxQuestionsPerIP) {
+      client.emit('max questions');
+      log('client with IP ' + ipAddress + ' already has max questions in queue');
     } else if (/\S/.test(data)) {
+      questionsInQueuePerIP[ipAddress]++;
       questions[client.id] = {question: data, id: client.id};
-      console.log(questions);
+      // log(questions);
       socket.emit('new question entered', Object.keys(questions).length);
     } else {
-      console.log('blank question');
+      log('blank question');
     }
   });
 
   client.on('clear question', function() {
     delete questions[client.id];
-    console.log('cleared question, ' + questions);
+    if (questionsInQueuePerIP[ipAddress]) {
+      questionsInQueuePerIP[ipAddress]--;
+    }
+    // log('cleared question, ' + questions);
   });
 
   client.on('submit answer', function(data) {
-    data = data.replace(/</g, "&lt;").replace(/>/g, "&gt;");
-    if (typeof answers[client.id] == "undefined") {
+    data = escapeHtml(data);
+    if (!answers[client.id]) {
```

Comment:
typeof undefined checks are good, but are almost always unnecessarily specific. One of the awesome things about JavaScript is its falsiness checks, e.g. `!variable` is almost always just as fine as a stricter check.